pkg/ring/replication_set_test.go

```diff
@@ -33,3 +37,233 @@ func TestReplicationSet_GetAddresses(t *testing.T) {
 		})
 	}
 }
+
+var (
+	errFailure     = errors.New("failed")
+	errZoneFailure = errors.New("zone failed")
+)
+
+// Return a function that fails starting from failAfter times
+func failingFunctionAfter(failAfter int, delay time.Duration) func(context.Context, *IngesterDesc) (interface{}, error) {
+	var mutex = &sync.RWMutex{}
+	count := 0
```

Comment:
using `atomic.Int32` can simplify the test a bit.


```suggestion
func failingFunctionAfter(failAfter int32, delay time.Duration) func(context.Context, *IngesterDesc) (interface{}, error) {
	count := atomic.NewInt32(0)
	return func(context.Context, *IngesterDesc) (interface{}, error) {
		time.Sleep(delay)
		if count.Inc() > failAfter {
			return nil, errFailure
		}
		return 1, nil
	}
}
```